core[minor]: Enhance cache flexibility in BaseChatModel

alhridoy opened the pull request:
Please title your PR "core: Enhance cache flexibility in BaseChatModel".

  - **Description:** Enhanced the `BaseChatModel` to support an `Optional[Union[bool, BaseCache]]` type for the `cache` attribute, allowing for both boolean flags and custom cache implementations. Implemented logic within chat model methods to utilize the provided custom cache implementation effectively. This change aims to provide more flexibility in caching strategies for chat models.
  - **Issue:** Implements enhancement request #17242.
  - **Dependencies:** No additional dependencies required for this change.

alhridoy committed:
Modified  to support  for the  attribute

Lord-Haji reviewed (changes requested):
Why are you making so many indent changes? The issue only specified cache. I'd suggest you to revert them

eyurtsev commented:
Hi @alhridoy, thanks for taking this on!!

Could you review the acceptance criteria in the original issue: https://github.com/langchain-ai/langchain/issues/17242 -- we want to implement all the items in the acceptance criteria so we can merge the PR.

eyurtsev commented:
We should iterate in smalls steps and match existing global cache functionality. If we want to extend the caching for streaming, we can do so, but let's do it as a separate task if it's not covered by existing global cache.

alhridoy committed:
added doctring in BaseChatModel and update unit tests

alhridoy commented:
> Hi @alhridoy, thanks for taking this on!!
> 
> Could you review the acceptance criteria in the original issue: #17242 -- we want to implement all the items in the acceptance criteria so we can merge the PR.

Hi @eyurtsev I added the inline docstring and added test. Could not work on external documentation. Could anybody help me here @hasansustcse13 @Lord-Haji

eyurtsev committed:
x

eyurtsev committed:
x

eyurtsev committed:
x

eyurtsev committed:
lint

eyurtsev commented on libs/core/langchain_core/language_models/chat_models.py:
```diff
@@ -103,8 +104,16 @@ async def agenerate_from_stream(
 
-    cache: Optional[bool] = None
-    """Whether to cache the response."""
+    cache: Optional[bool | BaseCache] = None
```
@alhridoy I made a few changes in this file. We should document the property itself.

We can ignore examples for now. (It would be nice to get documentation in the main python site, but we can do that afterwards.)

The main change was to streamline the caching path / non caching code paths, so there's only one path for caching related logic independently of whether a local cache is used or a global one

eyurtsev commented:
I'm going to commandeer to we can land this

alhridoy commented:
> I'm going to commandeer to we can land this

Hi @eyurtsev Give me a few days. I'll try it first. Thanks for your patience and guidance.

eyurtsev committed:
x

eyurtsev committed:
Merge branch 'master' into fork/cache

eyurtsev committed:
revert this test

eyurtsev committed:
Merge branch 'master' into fork/cache

eyurtsev committed:
x

eyurtsev committed:
x

eyurtsev commented:
@alhridoy just saw your message -- i made the changes already. I'm going to merge for now if tests pass, feel free to review the changes. And there are a lot of contribution opportunities so we'll appreciate any help :)

eyurtsev committed:
x

eyurtsev committed:
lint

eyurtsev committed:
Merge branch 'master' into cache

alhridoy reviewed (commented):
LGTM!

eyurtsev committed:
Merge branch 'master' into cache

eyurtsev committed:
x

eyurtsev committed:
x

eyurtsev committed:
x

eyurtsev committed:
x

eyurtsev committed:
x
